pkg/handler/v2/external_cluster/external_cluster.go

```diff
@@ -795,3 +797,50 @@ type body struct {
 	Kubeconfig string                          `json:"kubeconfig,omitempty"`
 	Cloud      *apiv2.ExternalClusterCloudSpec `json:"cloud,omitempty"`
 }
+
+func GetKubeconfigEndpoint(userInfoGetter provider.UserInfoGetter, projectProvider provider.ProjectProvider, privilegedProjectProvider provider.PrivilegedProjectProvider, clusterProvider provider.ExternalClusterProvider, privilegedClusterProvider provider.PrivilegedExternalClusterProvider, settingsProvider provider.SettingsProvider) endpoint.Endpoint {
+	return func(ctx context.Context, request interface{}) (interface{}, error) {
+		if !AreExternalClustersEnabled(settingsProvider) {
+			return nil, errors.New(http.StatusForbidden, "external cluster functionality is disabled")
+		}
+
+		req := request.(getClusterReq)
+		if err := req.Validate(); err != nil {
+			return nil, errors.NewBadRequest(err.Error())
+		}
+
+		project, err := common.GetProject(ctx, userInfoGetter, projectProvider, privilegedProjectProvider, req.ProjectID, nil)
+		if err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+
+		cluster, err := getCluster(ctx, userInfoGetter, clusterProvider, privilegedClusterProvider, project.Name, req.ClusterID)
+		if err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+
+		kubeconfigReference := cluster.Spec.KubeconfigReference
+		if kubeconfigReference == nil {
+			return nil, fmt.Errorf("kubeconfig not available for the Cluster")
+		}
+
+		secretKeyGetter := provider.SecretKeySelectorValueFuncFactory(context.Background(), privilegedClusterProvider.GetMasterClient())
+
+		rawKubeconfig, err := secretKeyGetter(kubeconfigReference, resources.KubeconfigSecretKey)
+		if err != nil {
+			return nil, err
+		}
+
+		kubeconfig, err := base64.StdEncoding.DecodeString(rawKubeconfig)
+		if err != nil {
+			panic(fmt.Sprintf("Invalid base64 string %q", rawKubeconfig))
+		}
+
+		kubeconfigYaml, err := yaml.JSONToYAML(kubeconfig)
+		if err != nil {
+			return nil, err
+		}
+
+		return string(kubeconfigYaml), nil
```

Comment:
you have to return this as a file: https://github.com/kubermatic/kubermatic/blob/master/pkg/handler/common/kubeconfig.go#L467

https://github.com/kubermatic/kubermatic/blob/master/pkg/handler/v2/routes_v2.go#L907